### PR TITLE
Replacement of flake8 with Ruff

### LINKED ISSUE #1969: Replace flake8 with Ruff for faster linting
#### The problem

Linting speed can be improved by using only Ruff.
#### Requested Solution

I propose replacing flake8 with Ruff, which we are already using in the project. Ruff's linter is much faster and implements every rule in flake8.

#### Alternatives

Other popular python linters, but Ruff seems to be the fastest and its already being used.

### PR DESCRIPTION
**Pull request by SAMAD101**
### Description

This PR removes flake8 from `.pre-commit-config.yaml` , `setup.cfg` and adds ruff for linting

### Related Issue

Resolves #1969 

### Motivation and Context

ruff is faster so yeah...

### How Has This Been Tested?

Manually tested, `pre-commit`, `make lint` works fine.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:

- [x] I have read the [CONTRIBUTING](https://vorta.borgbase.com/contributing/) guide.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

*I provide my contribution under the terms of the [license](./../LICENSE.txt) of this repository and I affirm the [Developer Certificate of Origin][dco].*

[dco]: https://developercertificate.org/

### CODE REVIEW

**Comment by shivansh02** [comment]
Screenshot from 2024-04-05 01-21-00

I tried checking E201 and E241 of flake8 but ruff couldn't catch them and were fixed by black.

**Comment by Hofer-Julian** [comment]
> I tried checking E201 and E241 of flake8 but ruff couldn't catch them and were fixed by black.

Both rules are marked as preview, so I assume they will only be enabled in the next ruff release: https://docs.astral.sh/ruff/rules/#error-e

**Comment by shivansh02** [comment]
> > I tried checking E201 and E241 of flake8 but ruff couldn't catch them and were fixed by black.
> 
> Both rules are marked as preview, so I assume they will only be enabled in the next ruff release: https://docs.astral.sh/ruff/rules/#error-e

That makes sense. But I tried some stable ones (E501 and E703), failed to catch them also. @SAMAD101 can you check if this is working for you?

**Comment by shivansh02** [comment]
Also since more than half of the pycodestyle rules are still in preview, should we enable ruff's preview mode or wait for them to get stable before trying to replace flake8?

**Comment by SAMAD101** [comment]
> I tried checking E201 and E241 of flake8 but ruff couldn't catch them and were fixed by black.

Might be because it had to configured in `ruff.toml` Works fine now. Checked

**Comment by SAMAD101** [comment]
> > > I tried checking E201 and E241 of flake8 but ruff couldn't catch them and were fixed by black.
> > 
> > 
> > Both rules are marked as preview, so I assume they will only be enabled in the next ruff release: https://docs.astral.sh/ruff/rules/#error-e
> 
> That makes sense. But I tried some stable ones (E501 and E703), failed to catch them also. @SAMAD101 can you check if this is working for you?

Those seem to be working fine as well.
You can check them using: `ruff check src/ --preview --select <rule_code>`

File: bin/.gitkeep
Comment (real-yfprojects) [inline bin/.gitkeep:1]: What's this folder used for?

File: ruff.toml
Comment (real-yfprojects) [inline ruff.toml:1]: Is it possible to keep all (ruff) configuration in one place? That is `pyproject.toml`.

**Review by real-yfprojects — COMMENTED** [review]
You should reset your fork's master brach to upstream since it is diverged for a long time already.

**Comment by SAMAD101** [comment]
> You should reset your fork's master brach to upstream since it is diverged for a long time already.

will do

File: setup.cfg [hunk 1]
```diff
@@ -72,16 +72,8 @@ source = vorta
...
-ignore =
-max-line-length = 120
-extend-ignore = E203,E121,E123,E126,E226,E24,E704,W503,W504
-exclude =
-    build,dist,.git,.idea,.cache,.tox,.eggs,
-    ./src/vorta/__init__.py,.direnv,env
-
 [tox:tox]
-envlist = py36,py37,py38,flake8
+envlist = py36,py37,py38,ruff
```
Comment (m3nu) [inline setup.cfg:76]: Does ruff need it's own Python env? Wasn't it a Rust binary?

**Comment by m3nu** [comment]
Just to be clear slow linting isn't a problem we have. So this change should also simplify the config in some way or make it easier to run.

File: setup.cfg (hunk 1 above)
Comment (m3nu) [reply setup.cfg:76]: As can be seen from those old Python versions, the tox config isn't maintained any more. So I'd just run ruff in pre-commit hooks which are the main mechanism for this right now.

**Comment by SAMAD101** [comment]
> Just to be clear slow linting isn't a problem we have. So this change should also simplify the config in some way or make it easier to run.

yeah, it does somewhat simplify the cofigs. And linting is just as easy to run as it was, we can use pre-commit or 'ruff check'

File: setup.cfg (hunk 1 above)
Comment (real-yfprojects) [reply setup.cfg]: If the tox is unmaintained and no longer used I guess we should remove in a seperate PR?

**Pull request merged**